### ISSUE TITLE
eduvpn-cli triggers (unnecessary) re-authentication in browser after every 1 or 2 reboots

### ISSUE THREAD

**Issue by Lukas-UAUX** [eduvpn/python-eduvpn-client#521]
# Installation
Via Ubuntu Repository

# Version
4.0.1

# OS/Distribution
Ubuntu 22.04

# Bug
I have configured eduvpn to connect directly after login - before 4.0 I just used `nmcli` to start the "eduVPN" connection which worked perfectly.

Now with 4.0/4.0.1 the connection gets removed from NM on disconnect - so I changed my command to use the eduvpn-cli like this:
```bash
eduvpn-cli connect -n 1
```
While this seemed to work fine I noticed that after every 1 or 2 reboot (with a day in between) instead of being connected a browser opens to ask for credentials again even though the certificate should have plenty of time remaining.

**Comment by jwijenbergh** [comment]
That is annoying.

Couple of questions:

- Does this also  happen with the GUI, or have you not tried?
- Could you find the entry in your keyring (with seahorse) and tell me what the "expires" field states (copy the password input and give me the value of the "expires" key in the json)
- Maybe change your command to `eduvpn-cli -d connect -n 1` so that we can get some more logging? 

Thanks

**Comment by Lukas-UAUX** [comment]
- Today at least it also asked again but I am not sure if it did the last time.
- In Seahorse I can see an eduvpn entry for OAuth Tokens for our server (although we do not use OAuth?).
- I've added the `-d` option and let the command write to a log file lets hope that will help

Maybe it has sth. to do with shutdown before gracefully stopping the connection? The NM-Connection is also left behind in this case.
The GUI-App would be closed on shutdown and would (kinda) gracefully end the connection but what happens when using the cli?

**Comment by jwijenbergh** [comment]
> Today at least it also asked again but I am not sure if it did the last time.

Right, the GUI and CLI use the same keyring storage so this should also not happen, so an annoying bug

> In Seahorse I can see an eduvpn entry for OAuth Tokens for our server (although we do not use OAuth?).

Every eduVPN server uses OAuth under the hood, this is the opening browser process. It uses those tokens to get the certificate, hence why they are in a keyring and not plaintext. If the tokens are invalid/expired it needs to re-authorize so it could be that we are loading/saving them incorrectly

> I've added the -d option and let the command write to a log file lets hope that will help

Thanks. Note that log is already saved at `~/.config/eduvpn/log` as well. However, we log only fatal errors (and some warnings) without providing the debug `-d` flag

> Maybe it has sth. to do with shutdown before gracefully stopping the connection? The NM-Connection is also left behind in this case.

It should update the tokens in the keyring right as it gets a new configuration/certificate from the server (so before connect), such that it can use these again when you relaunch the GUI/CLI.

> The GUI-App would be closed on shutdown and would (kinda) gracefully end the connection but what happens when using the cli?

You are safe to close the GUI on connect. It doesn't yet terminate the connection, see #518.

Also what is the output of `eduvpn-cli status` when you are connected? What is the expiry there?

**Comment by jwijenbergh** [comment]
By the way, as a workaround you can connect with eduvpn and only disconnect with networkmanager (without having the gui/cli running) which does not delete the connection.

**Comment by Lukas-UAUX** [comment]
```
2023-03-29 08:10:05,011 - <function ApplicationModelTransitions.get_previous_servers at 0x7f95c15e5a20> - DEBUG - eduvpn.app - app.py:42 - Transition: NO_SERVER, old state: DEREGISTERED
2023-03-29 08:10:05,011 - MainThread - DEBUG - eduvpn.nm - nm.py:762 - calling action with CLI mainloop
2023-03-29 08:10:05,016 - MainThread - DEBUG - eduvpn.nm - nm.py:766 - Quiting main loop, thanks!
2023-03-29 08:10:05,016 - MainThread - DEBUG - eduvpn.nm - nm.py:762 - calling action with CLI mainloop
2023-03-29 08:10:05,020 - <function ApplicationModelTransitions.loading_server at 0x7f95c15e5c60> - DEBUG - eduvpn.app - app.py:66 - Transition: LOADING_SERVER, old state: NO_SERVER
2023-03-29 08:10:05,021 - <function ApplicationModelTransitions.chosen_server at 0x7f95c15e5e10> - DEBUG - eduvpn.app - app.py:71 - Transition: CHOSEN_SERVER, old state: LOADING_SERVER
2023/03/29 08:10:05 - Go - DEBUG - Getting access token
2023/03/29 08:10:05 - Go - DEBUG - Refresh token is empty, returning error
2023-03-29 08:10:05,039 - <function ApplicationModelTransitions.start_oauth at 0x7f95c15e6680> - DEBUG - eduvpn.app - app.py:96 - Transition: OAUTH_STARTED, old state: CHOSEN_SERVER
2023-03-29 08:10:05,039 - open-browser - DEBUG - eduvpn.app - app.py:112 - Opening web browser with url: https://[SERVER-URL]
2023-03-29 08:10:10,063 - open-browser - DEBUG - eduvpn.app - app.py:116 - Running os.wait for browser
```

**Comment by fkooman** [comment]
The OAuth server in eduVPN 2.x server does not return a new `refresh_token` when using the refresh token for a new access token. The initially obtained refresh token remains valid. This is acceptable in OAuth 2.0, but no longer in 2.1 draft which the 3.x server implements.

**Comment by jwijenbergh** [comment]
Should be fixed in the latest version, if not please re-open
